### PR TITLE
add callbacks nodes to channels

### PR DESCRIPTION
**Pull request by quentinql**

### CODE REVIEW

File: pypeman/channels.py [hunk 1]
```diff
@@ -66,6 +67,10 @@ def __init__(self, name=None, parent_channel=None, loop=None, message_store_fact
         self._status = BaseChannel.STOPPED
         self.processed_msgs = 0
         self.interruptable_sleeper = Sleeper(loop)  # for interruptable sleeps
+        self.done_callback = None
```
Comment (feenes) [inline pypeman/channels.py:70]: please explain exactly when a callback will be called and whether they are mutually exclusive or whether it is possible that two or more callbacks can be called for the same node under certain circumstances.

File: pypeman/channels.py [hunk 2]
```diff
@@ -66,6 +67,10 @@ def __init__(self, name=None, parent_channel=None, loop=None, message_store_fact
...
+        self.fail_callback = None
+        self.drop_callback = None
+        self.reject_callback = None  # TODO: maybe add a 'default' callback that is launch everytime (may be useful for file cleaning for example)
```
Comment (feenes) [inline pypeman/channels.py:73]: that is lauched. Also comment is difinetely too long. so put it above the line

File: pypeman/channels.py [hunk 3]
```diff
@@ -318,16 +323,26 @@ async def handle(self, msg):
...
                 await self.message_store.change_message_state(msg_store_id, message.Message.PROCESSED)
+                if self.drop_callback:
+                    await self.drop_callback[0].handle(msg)
                 raise
             except Rejected:
                 await self.message_store.change_message_state(msg_store_id, message.Message.REJECTED)
+                if self.reject_callback:
+                    await self.reject_callback[0].handle(msg)
+                if self.fail_callback:
+                    await self.fail_callback[0].handle(msg)
```
Comment (feenes) [inline pypeman/channels.py:339]: so a rejection is a fail?

File: pypeman/channels.py [hunk 4]
```diff
@@ -480,6 +495,51 @@ def graph_dot(self, end=''):
             for entry in sub.graph_dot(end=end):
                 yield entry
 
+    def _init_callback_nodes(self, *callback_nodes):
```
Comment (feenes) [inline pypeman/channels.py:498]: doc string missing.

File: pypeman/channels.py [hunk 5]
```diff
@@ -480,6 +495,51 @@ def graph_dot(self, end=''):
...
+        callback_nodes = [node for node in flatten(callback_nodes) if node]
+        for node in callback_nodes:
+            node.channel = self
```
Comment (feenes) [inline pypeman/channels.py:501]: How can a function that is called init_callback_nodes() change channels of existing nodes?

File: pypeman/channels.py [hunk 6]
```diff
@@ -480,6 +495,51 @@ def graph_dot(self, end=''):
...
+
+        if len(callback_nodes) > 1:
+            previous_node = callback_nodes[0]
```
Comment (feenes) [inline pypeman/channels.py:504]: `> 1` or `> 0`

File: pypeman/channels.py [hunk 7]
```diff
@@ -480,6 +495,51 @@ def graph_dot(self, end=''):
...
+
+            for node in callback_nodes[1:]:
+                previous_node.next_node = node
+                previous_node = node
+        return callback_nodes
```
Comment (feenes) [inline pypeman/channels.py:509]: function name seems a little confusing.
it seems to do many things and not all are related to callbacks. but perhaps I'm missing some context

File: pypeman/channels.py [hunk 8]
```diff
@@ -480,6 +495,51 @@ def graph_dot(self, end=''):
...
+        """
+        Add a callback when channel ends correctly (and without raising Dropped)
+        """
+        if self.done_callback:
+            raise PypemanConfigError(f"done_callback already existing for channel {self.name}")
+        self.done_callback = self._init_callback_nodes(*callback_nodes)
+
+    def add_fail_callback(self, *callback_nodes):
+        """
+        Add a callback when channel raise an exception (except Dropped)
```
Comment (feenes) [inline pypeman/channels.py:521]: raises

File: pypeman/tests/test_channel.py [hunk 9]
```diff
@@ -74,6 +86,82 @@ def test_no_node_base_channel(self):
         self.start_channels()
         self.loop.run_until_complete(chan.handle(msg))
 
+    def test_done_callback(self):
+        """ Whether BaseChannel done_callback is working """
```
Comment (feenes) [inline pypeman/tests/test_channel.py:90]: you should also checked that it is not called in the cases where it should not be called.

File: pypeman/channels.py (hunk 1 above)
Comment (feenes) [reply pypeman/channels.py:70]: also document call order if multiple callbacks can be called

File: pypeman/channels.py [hunk 10]
```diff
@@ -480,6 +495,51 @@ def graph_dot(self, end=''):
...
+            previous_node = callback_nodes[0]
+
+            for node in callback_nodes[1:]:
+                previous_node.next_node = node
+                previous_node = node
+        return callback_nodes
+
+    def add_done_callback(self, *callback_nodes):
+        """
+        Add a callback when channel ends correctly (and without raising Dropped)
```
Comment (feenes) [inline pypeman/channels.py:513]: language is wrong.

We always add the callback.

perhaps something shorter to find but what's correct is:

"Add a callback, that will be called, when ...."

File: pypeman/tests/test_channel.py [hunk 11]
```diff
@@ -16,6 +16,18 @@
 from pypeman.tests.common import TstNode
 
 
+def raise_dropped(msg):
```
Comment (feenes) [inline pypeman/tests/test_channel.py:19]: doc string:
show how and why this function would be used instead of directly raising the exception.

**Review by feenes — CHANGES_REQUESTED** [review]
some minor feedback.

mostly documentation

File: pypeman/channels.py [hunk 12]
```diff
@@ -502,21 +593,48 @@ def reset_pypeman_channels():
 class SubChannel(BaseChannel):
     """ Subchannel used for forking channel processing. """
 
-    def callback(self, fut):
-        self.parent.sub_chan_tasks.remove(fut)
+    def _callback(self, fut):
+        """
+        """
+        ctx = contextvars.copy_context()
+        entrymsg = ctx.get(MSG_CTXVAR)
```
Comment (quentinql) [inline pypeman/channels.py:600]: reset()

File: pypeman/channels.py [hunk 13]
```diff
@@ -66,6 +70,12 @@ def __init__(self, name=None, parent_channel=None, loop=None, message_store_fact
         self._status = BaseChannel.STOPPED
         self.processed_msgs = 0
         self.interruptable_sleeper = Sleeper(loop)  # for interruptable sleeps
+        self.done_callback = None
+        self.fail_callback = None
+        self.drop_callback = None
+        self.reject_callback = None
+        # TODO: maybe add a 'default' callback that is launch everytime
```
Comment (quentinql) [inline pypeman/channels.py:77]: final_callback

File: pypeman/channels.py [hunk 14]
```diff
@@ -66,6 +70,12 @@ def __init__(self, name=None, parent_channel=None, loop=None, message_store_fact
         self._status = BaseChannel.STOPPED
         self.processed_msgs = 0
         self.interruptable_sleeper = Sleeper(loop)  # for interruptable sleeps
+        self.done_callback = None
```
Comment (quentinql) [inline pypeman/channels.py:73]: ok_callback

**Review by feenes — APPROVED** [review]

**Pull request merged**